implement for loops

aalok-t opened the pull request:
### Description

Add support to parse for loops of the form `for (iter in lower_bound..upper_bound) { loop_body }` by transforming it into a corresponding while loop.

### Test

Refer to tests in `move-compiler/transactional-tests/tests/control_flow/`

aalok-t commented on third_party/move/move-compiler/src/parser/syntax.rs:
```diff
@@ -1184,6 +1202,81 @@ fn parse_control_exp(context: &mut Context) -> Result<(Exp, bool), Box<Diagnosti
 
+// "for (iter = init; condition; update) loop_body" transforms into
+// "let iter = init; while (condition) { loop_body; update; }" and
+// "for (iter = init; condition; update; spec { ... }) loop_body" transforms into
```
Currently we support let `for (iter = init; { spec { ... }; condition }; update) loop_body` instead.

wrwg commented:
Just checking back, I think this is the version before we discussed in person?

aalok-t commented:
Yes, I will push the updates soon.

On Mon, Nov 6, 2023 at 9:51 PM Wolfgang Grieskamp ***@***.***>
wrote:

> Just checking back, I think this is the version before we discussed in
> person?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/aptos-labs/aptos-core/pull/10711#issuecomment-1797864848>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBPQRIETOOJZEDJ3HLD7LDDYDHD6FAVCNFSM6AAAAAA6TLUVHKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTOOJXHA3DIOBUHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

vineethk commented:
@aalok-t Can you let us know when we should review this PR? Maybe put a [not ready for review yet] or take down the PR in the meantime? Another alternative is to remove all reviewers and add them back again when this is ready for review.

vineethk commented:
I ran the following transactional test, and it returned `12`: not `10` as one would expect.

```
//# publish
module 0x42::m {
    fun main(): u64 {
        let y = 0;
        for (i in 0..10) {
            y = y + 1;
        };
        y
    }
}

//# run 0x42::m::main
```

I think we need to add more tests: 
- ones with nested for loops
- nested for + while,
- `for (_ in 0..10) {}` leads to an unexpected (to the user) error, 
- modifying the iterator variable has different semantics than rust, we should have test cases to show this behavior, and defend why it is okay to diverge from the rust behavior.

brmataptos commented:
Test I in loop

aalok-t commented on third_party/move/move-compiler/src/parser/syntax.rs:
```diff
@@ -1184,6 +1206,174 @@ fn parse_control_exp(context: &mut Context) -> Result<(Exp, bool), Box<Diagnosti
+    );
+
+    // To create the declaration "let flag = false", first create the variable flag, and then assign it to false
+    let flag_symb = Symbol::from("update_iter_flag");
```
To implement for loops at a parser level, we would need to include a flag (or some ghost variable) which cannot be referenced in the loop. One resolution is to create an error if the name of the ghost variable is referenced inside the loop.

aalok-t replied:
This is an inherent problem with having a flag. There are two ways to implement for loops without a flag. One is to convert it into a `loop`, which is a bit more involved, and the other is to add an if-statement for the loop body as done in the updated PR.

aalok-t commented:
> Test I in loop

The tests `for_loop_increment_iter` and `for_loop_non_terminating` update the iterator in the loop body.

aalok-t commented:
> I ran the following transactional test, and it returned `12`: not `10` as one would expect.
> 
> ```
> //# publish
> module 0x42::m {
>     fun main(): u64 {
>         let y = 0;
>         for (i in 0..10) {
>             y = y + 1;
>         };
>         y
>     }
> }
> 
> //# run 0x42::m::main
> ```
> 
> I think we need to add more tests:
> 
> * ones with nested for loops
> * nested for + while,
> * `for (_ in 0..10) {}` leads to an unexpected (to the user) error,
> * modifying the iterator variable has different semantics than rust, we should have test cases to show this behavior, and defend why it is okay to diverge from the rust behavior.

1. Fixed the error for the given program.
2. Added tests with nesting. 
3. The error for `for (_ in 0..10) {}` is `Invalid variable usage. Unbound variable '_'` as iterator is variable that is updated in the code. This is where we diverge from Rust. Do you suggest we add a clarifying error message for such a usage?
4. Unlike Rust, we do not consume the iterators in the loop and instead treat it like a mutable variable (as demonstrated in the updated tests).

brmataptos commented on third_party/move/move-compiler/src/parser/syntax.rs:
```diff
@@ -944,6 +946,16 @@ fn parse_term(context: &mut Context) -> Result<Exp, Box<Diagnostic>> {
+            if !ends_in_block || at_end_of_exp(context) {
+                return Ok(control_exp);
+            }
+            return parse_binop_exp(context, control_exp, /* min_prec */ 1);
```
I guess I wrote this code, but maybe it could use more docs.

aalok-t replied:
Fixed it. Thanks!

aalok-t commented on third_party/move/move-compiler-v2/tests/checking/typing/while_condition.move:
```diff
@@ -1,12 +1,11 @@
     fun t0() {
         while (true) ();
-        while (false) ()
+        while (false) ();
```
No, I just noticed the that it was non-uniform. It does not affect the test.

wrwg replied:
Actually it does, see the changed .exp file.

vineethk commented:
Consider the following program (which I would appreciate if you add as a test, this showcases the semantics not explored by tests so far, I think):

```
//# publish
module 0x42::m {

fun foo(x: &mut u64): u64 {
    *x = *x + 1;
    10
}

fun main(): u64 {
    let x = 0;
    for (i in 0..foo(&mut x)) {};
    x
}
}

//# run 0x42::m::main
```

We return `11` in this case: this is also a divergence from rust (which only calls `foo` once in an equivalent program), and might be surprising behavior for many (that the function gets evaluated multiple times, so for efficiency, they would need to manually create a temporary result).

So, I think we should discuss all the divergent (from rust) behaviors here to make sure we nail the semantics of a construct that will likely be used quite heavily. It is okay to diverge from other languages, but we need to have good arguments. It would be difficult to go back on some of these choices.

wrwg commented on third_party/move/move-compiler/src/parser/syntax.rs:
```diff
@@ -1184,6 +1206,127 @@ fn parse_control_exp(context: &mut Context) -> Result<(Exp, bool), Box<Diagnosti
+    let iter = parse_identifier(context)?;
+
+    assert!(
+        matches!(context.tokens.peek(), Tok::Identifier) && context.tokens.content() == IN_IDENT,
```
There is `consume_identifier` which does the check for the token kind and value, you can use it here and elsewhere.

wrwg reviewed (commented):
Looks good, nits only

brmataptos commented on third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move:
```diff
@@ -0,0 +1,20 @@
+                y = y - 1;
+            };
+        };
+        assert!(y == 1, 42);
```
This is an odd test case.  No matter how many times the first loop is traversed, y==1 here.

aalok-t replied:
Updated the test.

brmataptos commented on third_party/move/move-compiler/transactional-tests/tests/control_flow/nested_for_while_loops.move:
```diff
@@ -0,0 +1,20 @@
+        assert!(y == 1, 42);
+
+        while (y < 10) {
+            for (j in 0..2) {
```
And a lot of incorrect implementations of this for loop will still leave y==11 at the end.

aalok-t replied:
Updated the test.

brmataptos commented on third_party/move/move-compiler/transactional-tests/tests/control_flow/for_type_mismatch.exp:
```diff
@@ -0,0 +1,40 @@
+Error: error[E04007]: incompatible types
+  ┌─ TEMPFILE:5:9
+  │    
+5 │ ╭ ╭         for (i in true..false) {
```
This set of errors is very confusing to the user.  Can we somehow refine the set of errors we produce?

aalok-t replied:
Resolving this issue needs deeper changes. I will add it as an issue after this PR is merged.

aalok-t commented on third_party/move/move-compiler/transactional-tests/tests/control_flow/break_continue_for_loop.move:
```diff
@@ -0,0 +1,12 @@
+        let x = 0;
+        for (i in 0..10) {
+            if (x >= 5) break;
+            x = x + 1;
```
Added other tests where `i` occurs in loop body.

brmataptos commented:
Can we have a few tests for potential existing for usages, e.g.:
```
module 0x42::test {
  fun  for(i : u32, j: u32, k:u32) : u32 {
     i + j + k
  }
  fun for_user() : u32 {
      let (i, j, k) = (3, 4, 5);
      let x = for(i, j, k);
      x
  }
}
```

I see that `third_party/move/move-stdlib/nursery/sources/offer.move` has some more, so maybe building the stdlib as part of the normal tests will be useful, too.

brmataptos commented on third_party/move/move-compiler/src/parser/syntax.rs:
```diff
@@ -941,9 +943,15 @@ fn parse_term(context: &mut Context) -> Result<Exp, Box<Diagnostic>> {
+                && matches!(context.tokens.lookahead(), Ok(Tok::LParen)) =>
+        {
+            let (control_exp, _) = parse_for_loop(context)?;
+            return Ok(control_exp);
```
At least add a comment to say why this code isn't the same as the other control flow code.  perhaps:
```
for loop isn't useful in an expression today, so we ignore second result from `parse_for_loop` and never call `parse_binop_exp` (as might be done for some other control constructs above).
```

vineethk reviewed (commented):
@aalok-t: Did you end up addressing the two concerns here? I did not see corresponding test cases I provided added here, or a further discussion.

1. https://github.com/aptos-labs/aptos-core/pull/10711#issuecomment-1836788142
2. https://github.com/aptos-labs/aptos-core/pull/10711#issuecomment-1836803685

I just wanted to make sure we nail the semantics so that we are not stuck with supporting old behavior.

brmataptos commented on third_party/move/move-compiler/transactional-tests/tests/control_flow/break_continue_for_loop.exp:
```diff
@@ -0,0 +1,13 @@
+task 0 'run'. lines 1-12:
+Error: error[E01002]: unexpected token
+   ┌─ TEMPFILE:10:9
+   │
```
This is not a useful test.
